import-automation/progress-dashboard-rest/app/service/base_database.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Base class for a Database service that stores some kind of entities using
+Google Cloud Datastore for storage.
+"""
+
+from google.cloud import datastore
+
+from app import utils
+
+
+class BaseDatabase:
+    """Base class for a Database service that stores some kind of entities using
+    Google Cloud Datastore for storage.
+
+    New Entities must be created using get_by_id(make_new=True).
+
+    The datastore Client will attempt to infer credentials based on
+    host environment. See
+    https://cloud.google.com/docs/authentication/production#finding_credentials_automatically.
+
+    Attributes:
+        kind: Kind of entities to store
+        client: datastore Client to communicate with Datastore
+    """
+    def __init__(self, kind, client=None, id_field=None):
+        """Constructs an BaseDatabase.
+
+        Args:
+            kind: Kind of entities stored
+            client: Client to communicate with Datastore.
+            id_field: Name of the ID field of an entity as a string.
+                Every operation checks if id_field is not already a field in the
+                entity. If not, it will be added to the entity with the value of
+                the key name.
+        """
+        if not client:
+            client = utils.create_datastore_client()
+        self.kind = kind
+        self.client = client
+        self.id_field = id_field
+
+    def get_by_id(self, entity_id=None, make_new=False):
+        """Retrieves an entity from Datastore given its entity_id.
+
+        Args:
+            entity_id: ID of the entity as a string.
+            make_new: Whether to return a new entity if the entity_id does
+                not exist, as a boolean. If entity_id is None, make_new must
+                be True.
+
+        Returns:
+            The entity with the entity_id, as a datastore Entity. None,
+            if the entity_id does not exist and make_new is False.
+
+        Raises:
+            ValueError: entity_id is None and make_new is False
+        """
+        if not entity_id and not make_new:
```

Comment:
Is this equivalent?
```
  # Find an existing entity or make a new one.                                  
  entity = None
  if entity_id:
      entity = self.client.get(self._get_key(entity_id))
  if not entity	and make_new:
      entity = datastore.Entity(self._get_key(utils.get_id()))
  if entity:
      entity[self.id_field] = entity.key.name
      return entity
  # No entity was found or created.                                             
  if entity_id:
      # Entity not present.                                                     
      return None
  raise ValueError('entity_id is None and make_new is False')
```

Reply:
tests are failing so I think not

Reply:
get_by_id('does-not-exist', True) should return a new Entity, similar to dict.get. This code returns None